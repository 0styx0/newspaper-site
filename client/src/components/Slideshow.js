import React from 'react';
import { Link } from 'react-router-dom';

class Slideshow extends React.Component {

    constructor(props) {
        super(props);

        this.switchActiveImage = this.switchActiveImage.bind(this);

        this.state = {
            slideInfo: this.separateImages(this.props.images),
            images: [],
            activeImg: 0
        }
    }

    /**
     * Splits array of objects {img_url: json string array, slide_img: json string array of 0 or 1, issue: int, url: string}
     * into array of objects {img: image url, url: url of article}
     */
    separateImages(images) {

       const imageInfo = [];

       images.forEach((val) => {
            const img_url = JSON.parse(val.img_url)
            const display = JSON.parse(val.slide_img);


            img_url.filter((img, idx) => +display[idx] !== 0)
                    .forEach((img => imageInfo.push({
                        img,
                        url: `/issue/${val.issue}/story/${val.url}`
                    })))
        });

        if (imageInfo.length === 0) {

            imageInfo.push({
                img: "../images/tabc_logo.png",
                url: "/"
            });
        }

        return imageInfo;
    }

    componentWillMount() {

        const imagesWithLinks = this.state.slideInfo.map((img =>
                                  <Link to={img.url}>
                                    <img alt="" className="slideshowPic" src={img.img} />
                                  </Link>));

        this.setState({
            images: imagesWithLinks
        });

    }

    switchActiveImage(event) {

        this.setState({
            activeImg: this.state.activeImg + 1
        })

    }

    render() {


        if (!this.state.images[this.state.activeImg]) {
            return <span />
        }

        const imgClone = React.cloneElement(this.state.images[this.state.activeImg], {

           children: React.cloneElement(this.state.images[this.state.activeImg].props.children, {
               className: "slideshowPic activePic"
           }),

            onAnimationIteration: () => this.setState({
                activeImg: (this.state.activeImg + 1) % this.state.images.length
            })
        })


        return (

            <div id="slideShow">
                {imgClone}
            </div>
        );


    }
}

export default Slideshow;