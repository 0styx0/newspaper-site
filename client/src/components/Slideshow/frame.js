import React from 'react';
import PropTypes from 'prop-types';

import './index.css';


export default function Slideframe(props) {

    const imgClone = React.cloneElement(props.img, {

        children: React.cloneElement(props.img.props.children, {
            className: "slideshowPic activePic"
        }),

        onAnimationIteration: props.switchImg
    })

    return (

        <div id="slideShow">
            {imgClone}
        </div>
    );
}

Slideframe.propTypes = {

    img: PropTypes.element.isRequired,
    switchImg: PropTypes.func // required only if need more than 1 image in slideshow (so basically always)
}

