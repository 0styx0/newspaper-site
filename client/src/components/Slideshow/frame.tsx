import * as React from 'react';

import './index.css';

interface Props {
    img: JSX.Element;
    switchImg: Function // required only if need more than 1 image in slideshow (so basically always)
}

export default function Slideframe(props: Props) {

    const imgClone = React.cloneElement(props.img, {

        children: React.cloneElement(props.img.props.children, {
            className: "slideshowPic activePic"
        }),

        onAnimationIteration: props.switchImg
    })

    return (

        <div id="slideShow">
            {imgClone}
        </div>
    );
}
