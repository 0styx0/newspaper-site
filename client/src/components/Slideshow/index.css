.slideshowPic, #placeholderPic {
    max-width: 100%;
    max-height: 300px;
    width: 500px; /* width and height are also specified in js to overide inline styles */
    height: 320px;
    margin: 0 50px;
    border: 2px solid black;
}

#slideShow {

    position: relative;
    align-self: center;
    text-align: center;
}

#placeholderPic {
    display: inline;
}

@keyframes fade {
    0% {
        opacity: 0;
    }

    50% {
        opacity: 1;
    }
    100% {
        opacity: 0;
    }
}


.activePic {
    animation: fade 10s forwards infinite;
}

.slideshowPic:not(.activePic) {
    display: none;
}

@media (max-width: 1155px) {

	.slideshowPic {
        max-height: 100%;
	}
}
