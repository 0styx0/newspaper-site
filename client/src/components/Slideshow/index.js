import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import Slideframe from './frame';


class Slideshow extends React.Component {

    constructor(props) {
        super(props);


        this.switchActiveImg = this.switchActiveImg.bind(this);

        this.state = {
            images: this.props.images,
            activeImg: 0
        }
    }

    componentWillMount() {


        const imagesWithLinks = this.state.images.map((img =>
                                  <Link to={img.url}>
                                    <img alt="" className="slideshowPic" src={img.img} />
                                  </Link>));

        this.setState({
            images: imagesWithLinks
        });

    }

    switchActiveImg() {

        this.setState({
                activeImg: (this.state.activeImg + 1) % this.state.images.length
            })
    }

    render() {

        const defaultImg = <Link to="/">
                                <img alt="Default, TABC logo" className="slideshowPic" src="../images/tabc_logo.png" />
                            </Link>


        const currentImg = this.state.images[this.state.activeImg] || defaultImg

        return <Slideframe
                 img={currentImg}
                 switchImg={this.switchActiveImg}
               />

    }
}


Slideshow.propTypes = {
    images: PropTypes.arrayOf(PropTypes.shape({
              url: PropTypes.string,
              img: PropTypes.string
    }))
}

export default Slideshow;