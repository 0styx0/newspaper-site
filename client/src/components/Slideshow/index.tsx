import * as React from 'react';
import { Link } from 'react-router-dom';

import Slideframe from './frame';

interface Image {
    url: string;
    img: string;
}

interface Props {
    images: Image[];
}

interface State {
    images: JSX.Element[] | Image[];
    activeImg: number;
}

class Slideshow extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.switchActiveImg = this.switchActiveImg.bind(this);

        this.state = {
            images: this.props.images,
            activeImg: 0
        }
    }

    componentWillMount() {

        const imagesWithLinks: JSX.Element[] = (this.state.images as Image[]).map(img => {

                                  return (<Link to={img.url}>
                                            <img alt="" className="slideshowPic" src={img.img} />
                                          </Link>
                                  );
        });

        this.setState({
            images: imagesWithLinks
        });
    }

    switchActiveImg() {

        this.setState({
                activeImg: (this.state.activeImg + 1) % this.state.images.length
            })
    }

    render() {

        const defaultImg = <Link to="/">
                                <img alt="Default, TABC logo" className="slideshowPic" src="../images/tabc_logo.png" />
                            </Link>

        const currentImg = this.state.images[this.state.activeImg] || defaultImg

        return <Slideframe
                 img={currentImg as JSX.Element}
                 switchImg={this.switchActiveImg}
               />

    }
}

export default Slideshow;