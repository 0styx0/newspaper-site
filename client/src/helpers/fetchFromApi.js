
function fetchFromApi(url, method = "get", body = {}) {

    const options = {
        credentials: "include",
        method,
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify(body)
    }

    if (method.toLowerCase() === "get") {
        delete options.body;
    }

    return fetch(`http://localhost:3000/api/${url}`, options);
}

export default fetchFromApi;