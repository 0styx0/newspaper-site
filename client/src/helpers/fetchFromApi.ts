
function fetchFromApi(url: string, method = "get", body = {}) {

    const options = {
        credentials: "include" as "include",
        method,
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify(body)
    }

    if (method.toLowerCase() === "get") {
        delete options.body;
    }

    return fetch(`http://localhost:3000/api/${url}`, options);
}

export default fetchFromApi;