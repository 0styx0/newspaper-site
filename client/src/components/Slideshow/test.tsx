import * as React from 'react';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import Slideshow from './';
import { Image } from './';
import casual from '../../tests/casual.data';
import { setupComponent } from '../../tests/enzyme.helpers';

import { mount } from 'enzyme';


casual.define('images', () => {

   let amount = casual.randomPositive;
   const images: Image[] = [];

   while (amount-- > 0) {
       images.push({
           url: casual.url,
           img: casual.url
       });
   }

   return images;
});

type CustomCasual = typeof casual & {images: Image[]};
const customCasual = casual as CustomCasual;

describe('<Slideshow>', () => {

    function setup(images: Image[]) {

        return mount(
            <MemoryRouter>
                <Slideshow images={images} />
            </MemoryRouter>
        );
    }

    describe('snapshots', () => {

        function snap(images: Image[]) {

            const tree = renderer.create(
                <MemoryRouter>
                    <Slideshow images={images} />
                </MemoryRouter>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }

        it('renders correctly', () => {

            snap([{
                    url: 'test',
                    img: 'hello.png'
                }]);
        });
    });

    describe('#switchActiveImage', () => {

        it('switches to next image', () => {

            const images = customCasual.images;

            const wrapper = setup(images);
            const component = setupComponent(wrapper, Slideshow);

            expect(component.state.activeImg).toBe(0);

            component.switchActiveImg();

            expect(component.state.activeImg).toBe(1);
        });

        it('loops back to first img', () => {

            const images = customCasual.images;
            const wrapper = setup(images);
            const component = setupComponent(wrapper, Slideshow);

            component.state.activeImg = images.length - 1;

            component.switchActiveImg();

            expect(component.state.activeImg).toBe(0);
        });
    });
});
