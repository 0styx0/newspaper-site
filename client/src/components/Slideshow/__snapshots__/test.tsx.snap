// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slideshow> snapshots renders correctly 1`] = `
<div
  id="slideShow"
>
  <a
    href="/test"
    onAnimationIteration={[Function]}
    onClick={[Function]}
  >
    <img
      alt=""
      className="slideshowPic activePic"
      src="hello.png"
    />
  </a>
</div>
`;
